File: fortran/src/main/java/de/irt/jfor/I4.java
package de.irt.jfor;

/**
 * version:     $Revision$
 * created by:  dst
 * created on:  19.10.2017 13:52
 * modified by: $Author$
 * modified on: $Date$
 */
public class I4 implements Ref {

    public int v = 0;

    public I4(int v) {
        this.v = v;
    }

    public String toString() {
        return Integer.toString(v);
    }

    public static I4 of() {
        return new I4((short)0);
    }

    public static I4 of(int value) {
        return new I4((short)value);
    }
}

File: fortran/src/main/java/de/irt/kfor/I4.kt
package de.irt.kfor

/**
 * version:     $Revision$
 * created by:  dst
 * created on:  09.01.2019 09:27
 * modified by: $Author$
 * modified on: $Date$
 */
interface I4 {

    companion object {
        operator fun invoke(v : Int = 0) = IRef(v)
        fun arr(ni : Int) = Arr(ni)
        fun mat(ni : Int,nj : Int) = Mat(ni, nj)
        fun cub(ni : Int,nj : Int, nk : Int) = Cub(ni, nj, nk)
    }

    data class Arr(override val ni : Int) : IArr {
        val arr = IntArray(len)
        override operator fun get(i : Int) = arr[index(i)]
        override operator fun set(i : Int, v : Int) {
            arr[index(i)] = v
        }
        override fun allocate(ni : Int) = if(this.ni==ni) this else Arr(ni)
    }

    class Mat(override val ni : Int, override val nj : Int) : IMat {
        val arr = IntArray(len)
        override operator fun get(i : Int, j : Int) : Int = arr[index(i,j)]
        override operator fun set(i : Int, j : Int, v : Int) {
            arr[index(i,j)] = v
        }
        override fun allocate(ni : Int, nj : Int) = if(this.ni==ni&&this.nj==nj) this else Mat(ni, nj)
    }

    class Cub(override val ni : Int, override val nj : Int,  override val nk : Int) : ICub {
        val arr = IntArray(len)
        override operator fun get(i : Int, j : Int, k : Int) : Int = arr[index(i,j,k)]
        override operator fun set(i : Int, j : Int, k : Int, v : Int) {
            arr[index(i,j,k)] = v
        }
    }
}